controller/cmd/service-mirror/cluster_watcher.go

```diff
@@ -0,0 +1,610 @@
+			Name:     remotePort.Name,
+			Protocol: remotePort.Protocol,
+			Port:     gatewayPort,
+		})
+	}
+	return endpointsPorts
+}
+
+func (rcsw *RemoteClusterServiceWatcher) cleanupOrphanedServices() error {
+	matchLabels := map[string]string{
+		MirroredResourceLabel:  "true",
+		RemoteClusterNameLabel: rcsw.clusterName,
+	}
+
+	servicesOnLocalCluster, err := rcsw.localAPIClient.Svc().Lister().List(labels.Set(matchLabels).AsSelector())
+	if err != nil {
+		return fmt.Errorf("failed obtaining local services while GC-ing: %s", err)
+	}
+
+	for _, srv := range servicesOnLocalCluster {
+		_, err := rcsw.remoteAPIClient.Svc().Lister().Services(srv.Namespace).Get(rcsw.originalResourceName(srv.Name))
+		if err != nil {
+			// service does not exist anymore. Need to delete
+			if err := rcsw.localAPIClient.Client.CoreV1().Services(srv.Namespace).Delete(srv.Name, &metav1.DeleteOptions{}); err != nil {
+				rcsw.log.Errorf("Failed to GC local service %s", srv.Name)
+			} else {
+				rcsw.log.Debugf("Deleted service %s/%s as part of GC process", srv.Namespace, srv.Name)
+			}
+		}
+	}
+	return nil
+}
+
+// Whenever we stop watching a cluster, we need to cleanup everything that we have
+// created. This piece of code is responsible for doing just that. It takes care of
+// services, endpoints and namespaces (if needed)
+func (rcsw *RemoteClusterServiceWatcher) cleanupMirroredResources() error {
+	matchLabels := map[string]string{
+		MirroredResourceLabel:  "true",
+		RemoteClusterNameLabel: rcsw.clusterName,
+	}
+
+	services, err := rcsw.localAPIClient.Svc().Lister().List(labels.Set(matchLabels).AsSelector())
+	if err != nil {
+		return fmt.Errorf("could not retrieve mirrored services that need cleaning up: %s", err)
+	}
+
+	for _, svc := range services {
+		if err := rcsw.localAPIClient.Client.CoreV1().Services(svc.Namespace).Delete(svc.Name, &metav1.DeleteOptions{}); err != nil {
+			return fmt.Errorf("Could not delete  service %s/%s: %s", svc.Namespace, svc.Name, err)
+		}
+		rcsw.log.Debugf("Deleted service %s/%s", svc.Namespace, svc.Name)
+
+	}
+	edpoints, err := rcsw.localAPIClient.Endpoint().Lister().List(labels.Set(matchLabels).AsSelector())
+	if err != nil {
+		return fmt.Errorf("could not retrieve Endpoints that need cleaning up: %s", err)
+	}
+
+	for _, endpt := range edpoints {
+		if err := rcsw.localAPIClient.Client.CoreV1().Endpoints(endpt.Namespace).Delete(endpt.Name, &metav1.DeleteOptions{}); err != nil {
+			return fmt.Errorf("Could not delete  Endpoints %s/%s: %s", endpt.Namespace, endpt.Name, err)
+		}
+		rcsw.log.Debugf("Deleted Endpoints %s/%s", endpt.Namespace, endpt.Name)
+
+	}
+
+	namespaces, err := rcsw.localAPIClient.NS().Lister().List(labels.Set(matchLabels).AsSelector())
+	if err != nil {
+		return fmt.Errorf("could not retrieve Namespaces that need cleaning up: %s", err)
+	}
+
+	for _, ns := range namespaces {
+		if err := rcsw.localAPIClient.Client.CoreV1().Namespaces().Delete(ns.Name, &metav1.DeleteOptions{}); err != nil {
```

Comment:
Yes, was thinking about that as well. Do oyu know of the top off your head how you check if a namepsace is empty. Do you need to try all namespaced resources. Maybe will look into using the `dynamic` api

Reply:
Should we delete namespaces at all?

Reply:
It's a fair question.  If we create a namespace for a mirrored service, and then that service is deleted on the remote cluster, we would delete the local mirror.  This would leave the namespace we created lying around.  Not necessarily a problem, but seems a bit unhygienic.

Reply:
Yeah I am thinking we should be deleting them if they are empty. just could not figure out a good way to check for emptyness. but will find out.

Reply:
Please do the investigation, but it feels like checking for emptiness is going to be super fraught. We won't have RBAC to see everything and arbitrary CRs can be added to each namespace. I'd rather err on the side of being non-destructive and provide a tool to end users they can use to do cleanup manually (like `docker system prune`).